sdk/identity/azure-identity/inc/azure/identity/managed_identity_credential.hpp

```diff
@@ -102,7 +103,7 @@ namespace Azure { namespace Identity {
      * @param id The resource ID of the user-assigned managed identity.
      *
      */
-    static ManagedIdentityId FromUserAssignedResourceId(Azure::Core::ResourceIdentifier id)
+    static ManagedIdentityId FromUserAssignedResourceId(Azure::Core::ResourceIdentifier const& id)
```

Comment:
I don't feel confident we can safely make a change like this, and not have a need to potentially remove it in the future (which would be a breaking change).

I don't think we should only use the current (point-in-time) implementation when marking things like these as const, since implementations can change and we are changing the contractual guarantee here.

We should specify const only to places where we know that it can be done so by design, and won't expect it to ever change.

Reply:
We can safely change by-vaue and by-const-ref, it is not a breaking change. Everything that was compiling before would still compile. If in the future we change it back, it won't be a breaking change either.

Reply:
Today, we only use `id.ToString()`. Plus, the thing lives in `_detail` namespace.

Reply:
> If in the future we change it back, it won't be a breaking change either. 

Oh really?

So, in C++ is this case is different than adding `const` to a method, since we can't take that back without breaking?

Reply:
Would the same hold true if it was const (non-ref) param? As in, could we add const to a method param and remove it, without a breaking change even if we accepted the param by value?

Reply:
Yes, we can add and remove `const` from any by-val param. Treat such params as basically temporary variables inside function's implementation. Whether it is const or not, it does not affect the caller.

Reply:
FWIW, I believe that it *can* affect the caller's code generation. And that might affect customers that build shared objects from our libraries and don't recompile their source code to correspond to updated headers.

But we've made a strong statement that we expect that all customers recompile ALL their code when they update our libraries. So there is no *practical* negative effect.
 
This is one of the core advantages of shipping a source level distribution.

Reply:
C++ can be so different, and I love it. On one side, you have a possibility to design your API so that you are ABI compatible (at least between versions built by the same compiler) and could upgrade without recompiling. On the other side, there are header-only libraries, sometimes with non-trivial template metaprogramming.